### PR TITLE
Fix dependencies and https address of gitpython

### PR DESCRIPTION
**Pull request by JoshuaSBrown**
I was running into a checksum error with the old https address on pypi I have swapped it out and updated gitpython to use a newer version. The dependencies were also a little out of date. 

==> Error: ChecksumError: sha256 checksum failed for /tmp/joshuasbrown/spack-stage/spack-stage-py-gitpython-3.1.8-yar5zpp5q3hiamhpckkd6qphpwg2izzu/GitPython-3.1.8.tar.gz
    Expected 49d4f859bda7bd6ef9ce72d9a2cddb6a59dd70f34a18ce409670bc655431c517 but got 080bf8e2cf1a2b907634761c2eaefbe83b69930c94c66ad11b65a8252959f912

/home/joshuasbrown/Software/parthenon-project/spack/lib/spack/spack/package.py:1339, in do_fetch:
       1336        self.stage.fetch(mirror_only, err_msg=err_msg)
       1337        self._fetch_time = time.time() - start_time
       1338
  >>   1339        if checksum and self.version in self.versions:
       1340            self.stage.check()
       1341
       1342        self.stage.cache_local()

### CODE REVIEW

**Review by adamjstewart — CHANGES_REQUESTED** [review]
Can you keep the PyPI URL and instead fix the checksum for the 3.1.8 release? See https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#pypi-vs-github

**Comment by adamjstewart** [comment]
It looks like this package was added by @Sinan81 in #18387 and overwritten by @rvinaybharadwaj in #19194. I'm guessing someone accidentally used the checksum for the GitHub download instead of the PyPI download.

**Comment by JoshuaSBrown** [comment]
There is no tar file on pypi for 0.3.6 version https://pypi.org/project/GitPython/3.0.6/#files. So I don't know what to do about that.

**Comment by adamjstewart** [comment]
Just remove it

**Comment by JoshuaSBrown** [comment]
I think this solves the problem, I just used the github url for 0.3.6 release, the tar ball is hosted there, this should also help with backwards compatibility.

**Review by adamjstewart — CHANGES_REQUESTED** [review]
Please use `pypi = ` instead of `url = `, then you can skip the `url_for_version`, see https://spack.readthedocs.io/en/latest/build_systems/pythonpackage.html#pypi

**Comment by JoshuaSBrown** [comment]
If I do that and still want to include version 0.3.6 I get the error url referenced before assignment. When is the url populated if the pypi attribute is specified, is there a way to use urls from both github and pypi?

**Comment by adamjstewart** [comment]
You could do:
```python
version('0.3.6', sha256='e6599fcb939cb9b25a015a429702db39de10f2b493655ed5669c49c37707d233',
        url='https://github.com/gitpython-developers/GitPython/archive/0.3.6.tar.gz')
```
or simply delete the version since it's ancient.

File: var/spack/repos/builtin/packages/py-gitpython/package.py [hunk 1]
```diff
@@ -10,19 +10,23 @@ class PyGitpython(PythonPackage):
     """GitPython is a python library used to interact with Git repositories."""
 
     homepage = "http://gitpython.readthedocs.org"
-    pypi = "gitpython/GitPython-3.1.7.tar.gz"
+    pypi = "GitPython/GitPython-3.1.12.tar.gz"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-gitpython/package.py:13]:
```suggestion
    pypi = "gitpython/GitPython-3.1.12.tar.gz"
```
CamelCase doesn't work for me

File: var/spack/repos/builtin/packages/py-gitpython/package.py [hunk 2]
```diff
@@ -10,19 +10,23 @@ class PyGitpython(PythonPackage):
...
+    version('3.1.5',  sha256='90400ecfa87bac36ac75dfa7b62e83a02017b51759f6eef4494e4de775b2b4be')
+    version('3.1.4',  sha256='fa98ce1f05805d59bbc3adb16c0780e5ca43b5ea9422feecf1cd0949a61d947e')
+    version('3.1.3',  sha256='e107af4d873daed64648b4f4beb89f89f0cfbe3ef558fc7821ed2331c2f8da1a')
+    version('3.1.2',  sha256='864a47472548f3ba716ca202e034c1900f197c0fb3a08f641c20c3cafd15ed94')
+    version('3.1.1',  sha256='6d4f10e2aaad1864bb0f17ec06a2c2831534140e5883c350d58b4e85189dab74')
+    version('3.1.0',  sha256='e426c3b587bd58c482f0b7fe6145ff4ac7ae6c82673fc656f489719abca6f4cb')
+    version('3.0.9',  sha256='7e5df21bfef38505115ad92544fb379e6fedb2753f3b709174c4358cecd0cb97')
 
-    depends_on('python@3.4:',   type=('build', 'run'))
+    depends_on('python@3.5:',   type=('build', 'run'))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-gitpython/package.py]: 3.12 still supports Python 3.4, I would undo this change.

Comment (JoshuaSBrown) [reply var/spack/repos/builtin/packages/py-gitpython/package.py]: I was running into an error with setuptools:

```
spack install py-gitpython ^python@3.4
==> Warning: Failed to initialize repository: '/home/joshuasbrown/Software/kokkos-spack/kokkos'.
  No repo.yaml found in '/home/joshuasbrown/Software/kokkos-spack/kokkos'
  To remove the bad repository, run this command:
      spack repo rm /home/joshuasbrown/Software/kokkos-spack/kokkos
==> Error: An unsatisfiable version constraint has been detected for spec:

    python@3.4%gcc@7.4.0+bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib arch=linux-centos7-haswell
        ^bzip2
            ^diffutils
                ^iconv
        ^expat
        ^gdbm
            ^readline
                ^ncurses
                    ^pkgconf@1.7.3%gcc@7.4.0 arch=linux-centos7-haswell
        ^gettext+libxml2
            ^libxml2
                ^xz
                ^zlib@1.1.3:
        ^libffi
        ^openssl@:1.0.2z
            ^perl@5.14.0:
                ^berkeley-db
        ^sqlite@3.0.8:
        ^uuid


while trying to concretize the partial spec:

    py-setuptools@50.3.2%gcc@7.4.0 arch=linux-centos7-haswell


py-setuptools requires python version 3.5:, but spec asked for 3.4
```

Comment (JoshuaSBrown) [reply var/spack/repos/builtin/packages/py-gitpython/package.py]: Maybe I need to use an earlier version of setup tools for python3.4, I'm not quite sure how to specify that. I also would have thought that dependency should be resolved in the setuptools python package.py.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-gitpython/package.py]: You have a couple of options. Your best bet is to use the latest/default version of Python and not use 3.4 cuz it's ancient. If you really want to use 3.4, you can do:
```console
$ spack spec py-gitpython ^python@:3.4 ^py-setuptools@:43
```
The old concretizer isn't smart enough to handle this, but the new Clingo concretizer can handle it fine. You have to explicitly enable that at the moment though.

Comment (JoshuaSBrown) [reply var/spack/repos/builtin/packages/py-gitpython/package.py]: ok.

**Review by adamjstewart — APPROVED** [review]
